### ISSUE TITLE
Jit and gil

### ISSUE THREAD

**Issue by Yanpas** [RustPython/RustPython#362]
One of the reasons why cpython is the slowest one is that it is interpreted. Do you have plans to support jit? And will you keep gil?

**Comment by vokar97** [comment]
Speaking about JIT, there is already the existing issue: #300.

**Comment by cthulahoops** [comment]
On the subject of the GIL, ideally it would be nice to build a GIL-less python, but it's not something we've done any work on yet.

My expectation is that once we add the threading library Rust will start screaming that nothing is thread safe, and we'll have to add locks to make it run. I think we'll be aiming to add fine grained locks and see how it goes. The experience of CPython is that it's difficult to add more fine grained locking without breaking single core performance. Rust's safety checks might make this an easier task than in C, but that's just speculation at this stage.

And then there's the question of C extensions that expect to be able to acquire the GIL, so we may need an optional GIL for compatibility anyway.

In short, it's a very interesting can of worms without an easy answer.

**Comment by Yanpas** [comment]
As any regular user I'm not that interested what language an interpreter is written in. The whole point of switching to another interpreter (like pypy) is it's speed and/or concurrency (jython). So for the project to be successful I think it should have great compatibility with cpython and offer other benefits, even if it means that more code will be placed to unsafe blocks.

**Comment by Jacob-Riddle** [comment]
PyPy has experimented with [Transactional Memory](http://doc.pypy.org/en/latest/stm.html). Given the young nature of RustPython, I think some of the challenges PyPy has run into could be avoided.

**Comment by Ironlenny** [comment]
PyPy has experimented with [Transactional Memory](http://doc.pypy.org/en/latest/stm.html). Given the young nature of RustPython, I think some of the challenges PyPy has run into could be avoided.

**Comment by Mec-iS** [comment]
I suppose the major possible improvement that Rust can produce will be based on its built-in concurrency. As far as I can understand, this project is in the early phase in which the first step to a better Python interpreter is to clone CPython standard library in Rust and keep them in sync as much as possible. After that it will be possible to unleash all the advantages that Rust can bring in.

**Comment by jmfrank63** [comment]
I am rather new to the subject, but of course a GIL less python that is capable of multi threading would be of a great advantage. The way CPython is implemented makes it almost impossible to remove the GIL any attempt let to a huge amount of extra logs distributed all over different parts of the implementation slowing everything massively down.
However there had been successful implementations like the implementation of IronPython. So I would first study successful approaches instead of reinventing the wheel.
Admittedly IronPython is almost dead and still is missing Python 3 support, but the way the implementation had been done in respect to multithreading might be a used as a valuable template.
https://wiki.python.org/moin/IronPython

**Comment by youknowone** [comment]
basic jit with cranelift is recently added

**Comment by GoodiesHQ** [comment]
> 
> 
> I am rather new to the subject, but of course a GIL less python that is capable of multi threading would be of a great advantage. The way CPython is implemented makes it almost impossible to remove the GIL any attempt let to a huge amount of extra logs distributed all over different parts of the implementation slowing everything massively down.
> However there had been successful implementations like the implementation of IronPython. So I would first study successful approaches instead of reinventing the wheel.
> Admittedly IronPython is almost dead and still is missing Python 3 support, but the way the implementation had been done in respect to multithreading might be a used as a valuable template.
> https://wiki.python.org/moin/IronPython

Also, Cython is solidly mature at this point and creating functions that are C/Cython-only (any code that does not use the Python C API) can have the `nogil` attribute that will allow you to run parallelized code. Granted, you cannot use the Python API within that scope so python objects of any kind are off-limits... still worth a note, though! :)

**Comment by pca006132** [comment]
It seems that the current implementation uses OS thread for threading, but uses a lot of atomics for synchronization. I wonder if it is possible to have some kind of memory model that requires the user to explicitly synchronize the data, similar to using synchronize and volatile in Java. AFAIK, this is not defined for Python.

As mentioned in #2445 and #2474, atomic is causing serious performance issues, and the running time of simple programs increase linearly as the number of threads. We should try to minimize the synchronization needed for executing the code.

**Comment by Hypnoes** [comment]
According to this: [PEP 703](https://peps.python.org/pep-0703/) ;
Add a flag to describe code behaviors might be a good choice.
